LINVAST.Imperative/Builders/Java/JavaASTBuilder.Types.cs

```diff
@@ -14,6 +14,193 @@ namespace LINVAST.Imperative.Builders.Java
+            }
+
+            if (ctx.classDeclaration() is { }) {
+                var classDecl = this.Visit(ctx.classDeclaration()).As<TypeDeclNode>();
+                int declSpecsLine = ctxStartLine;
+                return new ClassNode(ctx.Start.Line, new DeclSpecsNode(declSpecsLine, modifiers, classDecl.Identifier), classDecl);
+            }
+
+            if (ctx.enumDeclaration() is { }) {
+                var enumDecl = this.Visit(ctx.enumDeclaration()).As<EnumDeclNode>();
+                int declSpecsLine = ctxStartLine;
+                return new EnumNode(ctx.Start.Line, new DeclSpecsNode(declSpecsLine, modifiers, enumDecl.Identifier), enumDecl);
+            }
+
+            if (ctx.interfaceDeclaration() is { }) {
+                var interfaceDecl = this.Visit(ctx.interfaceDeclaration()).As<TypeDeclNode>();
+                int declSpecsLine = ctxStartLine;
+                return new InterfaceNode(ctx.Start.Line, new DeclSpecsNode(declSpecsLine, modifiers, interfaceDecl.Identifier), interfaceDecl);
+            }
+            return new EmptyStatNode(ctx.Start.Line);
+        }
+
+        public override ASTNode VisitClassOrInterfaceModifier([NotNull] ClassOrInterfaceModifierContext ctx)
+        {
+
+            if (ctx.annotation() is { })
+                throw new NotImplementedException("annotations");
+
+            return new DeclSpecsNode(ctx.Start.Line, ctx.children.First().GetText());
+
+        }
+
+        public override ASTNode VisitTypeType([NotNull] TypeTypeContext ctx)
+        {
+
+            if (ctx.annotation() is { } && ctx.annotation().Any()) {
+                throw new NotImplementedException("annotations");
+            }
+
+            if (ctx.primitiveType() is { }) {
+                return this.Visit(ctx.primitiveType());
+
+            }
+
+            return this.Visit(ctx.classOrInterfaceType());
+
+        }
+
+
+        public override ASTNode VisitTypeList([NotNull] TypeListContext ctx)
+        {
+            var typeNameNodes = ctx.typeType().Select(c => this.Visit(c).As<TypeNameNode>()).ToArray<TypeNameNode>();
+
+
+            return new TypeNameListNode(ctx.Start.Line, typeNameNodes);
+        }
+
+        public override ASTNode VisitTypeParameters([NotNull] TypeParametersContext ctx)
+        {
+            return new TypeNameListNode(ctx.Start.Line, ctx.typeParameter().Select(c => this.Visit(c).As<TypeNameNode>()).ToArray());
+        }
+
+        public override ASTNode VisitTypeParameter([NotNull] TypeParameterContext ctx)
+        {
+            if (ctx.annotation() is { } && ctx.annotation().Any())
+                throw new NotImplementedException("annotations");
+
+            TypeNameListNode baseList;
+            if (ctx.typeBound() is { })
+                baseList = this.Visit(ctx.typeBound()).As<TypeNameListNode>();
+            else
+                baseList = new TypeNameListNode(ctx.Start.Line, new ArrayList<TypeNameNode>());
```

Comment:
You can just use `new TypeNameNodeList(ctx.Start.Line)` in this case, as there is a constructor defined that takes `params TypeNameNode[]`. This also applies to other appearances of the same pattern.

Reply:
_new ArrayList<TypeNameNode>()_ is called several times in code, I removed every occurence